subsys/bluetooth/host/hci_core.c

```diff
@@ -853,41 +853,41 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
 	}
 #endif
 
-	if (IS_ENABLED(CONFIG_BT_CENTRAL)) {
-		if (evt->status) {
-			/*
-			 * if there was an error we are only interested in pending
-			 * connection so there is no need to check ID address as
-			 * only one connection can be in that state
-			 *
-			 * Depending on error code address might not be valid anyway.
-			 */
-			conn = find_pending_connect(NULL);
-			if (!conn) {
-				return;
-			}
-
-			conn->err = evt->status;
-
+	if (evt->status) {
 		/*
-		 * Handle advertising timeout after high duty directed
-		 * advertising.
+		 * If there was an error we are only interested in pending
+		 * connection. There is no need to check ID address as
+		 * only one connection can be in that state.
+		 *
+		 * Depending on error code address might not be valid anyway.
 		 */
-		if (conn->err == BT_HCI_ERR_ADV_TIMEOUT) {
-			atomic_clear_bit(bt_dev.flags, BT_DEV_ADVERTISING);
+		conn = find_pending_connect(NULL);
+		if (!conn) {
+			return;
 		}
 
-			bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
+		conn->err = evt->status;
 
-			/* check if device is market for auto connect */
+		if (IS_ENABLED(CONFIG_BT_PERIPHERAL)) {
+			/*
+			 * Handle advertising timeout after high duty directed
+			 * advertising.
+			 */
+			if (conn->err == BT_HCI_ERR_ADV_TIMEOUT) {
+				atomic_clear_bit(bt_dev.flags, BT_DEV_ADVERTISING);
+			}
+		}
+
+		if (IS_ENABLED(CONFIG_BT_CENTRAL)) {
+			/* Check if device is marked for auto connect. */
 			if (atomic_test_bit(conn->flags, BT_CONN_AUTO_CONNECT)) {
 				bt_conn_set_state(conn, BT_CONN_CONNECT_SCAN);
 			}
-
-			goto done;
 		}
-	} else {
-		BT_ASSERT(evt->status == 0);
+
+		bt_conn_set_state(conn, BT_CONN_DISCONNECTED);
```

Comment:
Is it ok to switch states back-to-back from BT_CONN_CONNECT_SCAN to BT_CONN_DISCONNECTED?

Reply:
how can I know? :)
CO opinion is needed

Reply:
It seems that you shouldn't be doing it, i.e. leave it in CONNECT_SCAN if that was set. It's e.g. what commit 861bae22b80 was trying to fix. Based on that commit it would also be ok to first transition to DISCONNECTED and after that to CONNECT_SCAN. @sjanc it'd be good to get your review of this detail as well since you're the author of the commit in question.

Reply:
state set removed

Reply:
@pdunaj the state change to DISCONNECTED still needs to be there though. I'm not sure if it needs to be there first always, or only for the case where we don't go to CONNECT_SCAN. However, if you want to be in line with what the code used to do, then in case of a non-zero status the state should always first go to to DISCONNECTEd. @sjanc 	did you have a chance to look at this yet?

Reply:
Also, it'd be nice if someone actually tested this in practice so we're not just talking theoretical stuff here :)